### ISSUE TITLE
continue training with interrupted model

### ISSUE THREAD

**Issue by pure-rgb** [keras-team/keras#22010]
Say, you have 

```python
lr_schedule = keras.optimizers.schedules.ExponentialDecay(
    initial_learning_rate=1e-2,
    decay_steps=10000,
    decay_rate=0.9
)
model_check = keras.callbacks.ModelCheckpoint('my_model.weights.h5')
optimizer = keras.optimizers.SGD(learning_rate=lr_schedule)
model.compile(optimizer )
model.fit(x_train, y_train, callbacks=[reduce_lr, model_check], epochs=50)
```

Now, say the training got interrupted in the middle of epoch 27. How would you load the model and continue training properly, considering the continued updated `lr` at epoch 27.

**Comment by Sikandar1310291** [comment]
## Solution

In **Keras 3**, when you use `ModelCheckpoint` with `save_weights_only=True` (the default when saving to `.weights.h5`), the optimizer state (including `iterations` which drives learning rate schedules) **is automatically saved** in the weights file.

To resume training correctly:

```python
import keras

# 1. Define the SAME schedule as before
lr_schedule = keras.optimizers.schedules.ExponentialDecay(
    initial_learning_rate=1e-2,
    decay_steps=10000,
    decay_rate=0.9
)

# 2. Create a fresh model
model = create_your_model()  # Same architecture as before

# 3. Compile BEFORE loading weights (this is crucial!)
optimizer = keras.optimizers.SGD(learning_rate=lr_schedule)
model.compile(optimizer=optimizer, loss='your_loss')

# 4. Load the weights - optimizer state is restored automatically
model.load_weights('my_model.weights.h5')

# 5. Continue training with initial_epoch set to where you left off
model.fit(
    x_train, y_train,
    epochs=50,
    initial_epoch=27,  # Start from epoch 27
    callbacks=[model_check]
)
```

### Key Points:
1. **Compile before load_weights()** - This instantiates the optimizer so its state can be restored.
2. **Use initial_epoch** - This ensures epoch numbering continues correctly in logs/callbacks.
3. **Same schedule definition** - The schedule object must match the original.

I've also submitted a PR (#22014) to add this guidance to the official Keras documentation.

**Comment by pure-rgb** [comment]
@Sikandar1310291 

How would we ensure the optimizer state load properly? Isn't .h5 only store model param?

What about .keras format. Does it hold only model pram or optimizer param both?

What about [keras.callbacks.BackupAndRestore](https://keras.io/api/callbacks/backup_and_restore/) - See [this ](https://keras.io/getting_started/faq/#:~:text=a%20lower%20loss.-,How%20can%20I%20ensure%20my%20training%20run%20can%20recover%20from%20program%20interruptions%3F,-To%20ensure%20the)

> To ensure the ability to recover from an interrupted training run at any time (fault tolerance), you should use a [keras.callbacks.BackupAndRestore](https://keras.io/api/callbacks/backup_and_restore#backupandrestore-class) callback that regularly saves your training progress, including the epoch number and weights, to disk, and loads it the next time you call Model.fit().
import numpy as np
import keras
class InterruptingCallback(keras.callbacks.Callback):
  """A callback to intentionally introduce interruption to training."""
  def on_epoch_end(self, epoch, log=None):
    if epoch == 15:
      raise RuntimeError('Interruption')
model = keras.Sequential([keras.layers.Dense(10)])
optimizer = keras.optimizers.SGD()
model.compile(optimizer, loss="mse")
x = np.random.random((24, 10))
y = np.random.random((24,))
backup_callback = keras.callbacks.experimental.BackupAndRestore(
    backup_dir='/tmp/backup')
try:
  model.fit(x, y, epochs=20, steps_per_epoch=5,
            callbacks=[backup_callback, InterruptingCallback()])
except RuntimeError:
  print('***Handling interruption***')
> # This continues at the epoch where it left off.
>  model.fit(x, y, epochs=20, steps_per_epoch=5,
            callbacks=[backup_callback])

**Comment by pure-rgb** [comment]
@Sikandar1310291 

> In Keras 3, when you use ModelCheckpoint with save_weights_only=True (the default when saving to .weights.h5), the optimizer state (including iterations which drives learning rate schedules) is automatically saved in the weights file.

I think this is not right. I just check

```
model = create_your_model()

model.compile(...)

model.optimizer.learning_rate # X

model.load_weights(...)

model.optimizer.learning_rate # same X
```

**Comment by dhantule** [comment]
Hi @pure-rgb,

I tried using [keras.callbacks.BackupAndRestore](https://keras.io/api/callbacks/backup_and_restore/), and it seems to handle this correctly, the model resumes training from where it left off after an interruption. Please refer this attached [gist](https://colab.sandbox.google.com/gist/dhantule/7125dc41ea570f54adf93db1323fe6f9/continue-training-with-interrupted-model-22010.ipynb).

**Comment by pure-rgb** [comment]
@dhantule 
The backup callback only saves 

```
{"epoch": 24, "batch": 0}
```

No optimizer state information are stored.

Consider this case: You train a model with epoch 50. You use warmup learninge rate and various scheduler setup. You also use backup callback. Now, your training interrupted at epoch 27. Now, you restarted the environment - then import modules, build model and compile again. Can the previous backup resume training from epoch 27 with last check optimizer state? If you say, with this backup, internally the keras will compute last check optimizer state and resume training, can we verify this.

**Comment by dhantule** [comment]
Hi @pure-rgb, 

BackupAndRestore does store the full model and optimizer variables to the backup directory.
I verified this by using a custom callback to check the optimizer's internal state immediately after a resume, please refer this attached [gist](https://colab.sandbox.google.com/gist/dhantule/7125dc41ea570f54adf93db1323fe6f9/continue-training-with-interrupted-model-22010.ipynb#scrollTo=wim-n2ReoZwG).

**Comment by innat** [comment]
I don't use this backup callback, as it takes time to save checkpoint, increasing the training time. I use this, probably much simpler. 

---

Say, we set 500 epoch and saving checkpoint every 5 epochs. Now, the training got interrupted after epoch 350 before finising 351. In that case, we can resume training as follows:

```python
model = Model()
model.load_weights(...) # last saved checkpoint

# optim, lr sched etc, same as before, i.e.,
steps_per_epoch = num_samples // batch_size
lr_schedule = CosineDecay(
    initial_learning_rate=1e-6,
    decay_steps=decay_steps,
    warmup_target=min(3e-4, 1e-4 * (batch_size / 2)),
    warmup_steps=warmup_steps,
    alpha=0.1,
)
optim = keras.optimizers.AdamW(
    learning_rate=lr_schedule,
    weight_decay=1e-5,
)

# Important
already_trained_steps = 350 * steps_per_epoch
optim.iterations.assign(already_trained_steps)

model.fit(
    ...
    initial_epoch=350,
)
```
